### ISSUE TITLE
Possibility to not install Singularity legacy files (symlinks etc.)

### ISSUE THREAD

**Issue by olifre** [apptainer/apptainer#1759]
### Version of Apptainer

What version of Apptainer (or Singularity) are you using? Run

```
apptainer version 1.2.4
```

### Expected behavior

It would be great to have an official build option to drop `singularity` legacy, e.g. the symlink, manpage etc. 

### Actual behavior

Symlink is always installed. 

Note that this causes issues downtstream — for example, the binary collides with another project carrying the name [`singularity`](http://www.emhsoft.com/singularity/), predating the container runtime, causing the few distros packaging Apptainer to have blockers of the packages against each other, e.g.:
https://bugs.gentoo.org/912200
Official support to drop the legacy / compat binary would help distros to drop this blocker, and potentially pave the way towards dropping the legacy links at some point (and also, ease packaging).

**Comment by GodloveD** [comment]
I think the symlink was originally created to ease transition from Singularity to Apptainer after the rename.  We would probably want to see that continue as the default behavior, but I don't see any reason not to add an option to omit the link.

**Comment by DrDaveD** [comment]
Dropping the alias won't be enough to be able to simultaneously install singularity and apptainer into the same prefix.  In particular, the run-singularity binary is needed to directly execute a container with one or the other.  That conflict will never go away.  The only way to have them both on the same machine is to put them in different prefixes.

**Comment by olifre** [comment]
> Dropping the alias won't be enough to be able to simultaneously install singularity and apptainer into the same prefix.

The game `singularity` (which predates the container runtime `singularity`) does not use any `run-singularity` binary (see my source link above), so installing this in parallel in the same prefix would in fact be possible if the `singularity` symlink would be dropped. This is what is causing blockers in distros, as the downstream bug report above illustrates. 

It's clear to me that the container runtimes `singularity` and `apptainer` can not really coexist in the same prefix — but that was not what the actual issue here was about  ;-).

**Comment by DrDaveD** [comment]
Why remove the alias if they can't both be installed anyway?  I don't understand that.  What is the issue about then?  I don't really understand the gentoo terminology so I don't know what is meant by "blocker" if it doesn't mean that you can't install both of them at the same time.

**Comment by DrDaveD** [comment]
Oh ok I glossed over your reference to a game.  Still, is that likely to be something that someone wants to install at the same time as the container program?

**Comment by olifre** [comment]
> Why remove the alias if they can't both be installed anyway?

They could, if the alias was not there. And my understanding up to now was that the alias was kept for legacy reasons only. 

> [...] so I don't know what is meant by "blocker" if it doesn't mean that you can't install both of them at the same time.

Your understanding is correct. The "blocker" (or `Conflicts` in Debian speak) has to be added by downstream distros since both packages install `/usr/bin/singularity`. In general, it's considered good practice for packages to avoid using naming similar to other packages for their libraries and binaries. 

As you may be aware, Debian has packaged `singularity` (the game) as `singularity` (similar as for example on ArchLinux), and the container runtime as `singularity-container` (as it is something which appeared much later). 

> Still, is that likely to be something that someone wants to install at the same time as the container program?

It is apparently likely enough that several downstream distros packaging both had to take care of a blocker / `Conflicts` in their packaging (see for example also Debian's packaging of `singularity-container`). 

To give some numbers, the game has over 0.2 million installations on Debian systems according to [popcon](https://qa.debian.org/popcon.php?package=singularity). I don't consider it unlikely that there are users amongst these who want to use Apptainer to build containers on their desktops, and they will experience an error due to the file conflict introduced by the legacy symlink. So basically, my question was whether the non-optional installation of the legacy symlink is important enough to warrant extra hassle for distro packagers and users.

**Comment by DrDaveD** [comment]
I think it's far too soon to remove the default of including the singularity symlink by default on a particular distro, but I wouldn't object to it using /etc/alternatives, if somebody would want to do that packaging effort.

**Comment by olifre** [comment]
> I think it's far too soon to remove the default of including the singularity symlink on a particular distro, but I wouldn't object to it using /etc/alternatives, if somebody would want to do that packaging effort.

Thanks for reopening! 

I agree with your first statement (which is why I did not suggest this in my initial issue report, as @GodloveD confirmed — no change to defaults was intended). 

Since `/etc/alternatives` is distro-specific (it is not present on anything EL, Gentoo, ArchLinux etc. to my knowledge), my proposal was to add a build option first, which allows to drop installation of the `singularity` symlink. 

In other words, I am asking for something like `singulairty-compat` which is on by default but can be turned off. 

This would:
* Allow distros to decide (and for source-based distros, the user can decide). Since distros usually stay with upstream's choice of defaults, again, this will not change the downstream experience rightaway. 
* Allow users to decide when they compile themselves. 
* Easy things for those rolling out local binary packages (if they want to drop the symlink). 
* Allow projects which depend on the container runtime `apptainer` to actually compile and install `apptainer` in their CI without the symlink and run their test suite, uncovering potential legacy uses of the symlink in their code.

**Comment by DrDaveD** [comment]
I don't want to encourage distros to leave out the symlink completely, since I think that's too early, and I don't see much advantage to having a compile option if the packaging has to go through the trouble of setting up /etc/alternatives anyway.  The bottom line is I think that this should be a packaging issue only.  The Apptainer source code does include packaging information for rpm and deb, so I think the way to move forward is to update the deb packaging included with Apptainer.  That could then be used as an example for other distros like gentoo.

**Comment by DrDaveD** [comment]
Ok, then please go ahead and make another issue if you would like to see that.

**Comment by olifre** [comment]
Thanks! 

For future reference, in case someone stumbles upon this: Since I believe `/etc/alternatives` will not solve any known issue (but increase packaging fragmentation / distro specifics instead), I will not open another issue myself requesting this. 

The original request (adding a compile-time option to solve the issues outlined in my [earlier comment](https://github.com/apptainer/apptainer/issues/1759#issuecomment-1780001992) ) is denied.
